pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,465 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslicemirroring
+
+import (
+	"fmt"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	discoveryinformers "k8s.io/client-go/informers/discovery/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	discoverylisters "k8s.io/client-go/listers/discovery/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller"
+	mirroringmetrics "k8s.io/kubernetes/pkg/controller/endpointslicemirroring/metrics"
+)
+
+const (
+	// maxRetries is the number of times an Endpoints resource will be retried
+	// before it is dropped out of the queue. Any sync error, such as a failure
+	// to create or update an EndpointSlice could trigger a retry. With the
+	// current rate-limiter in use (1s*2^(numRetries-1)) the following numbers
+	// represent the sequence of delays between successive queuings of an
+	// Endpoints resource.
+	//
+	// 1s, 2s, 4s, 8s, 16s, 32s, 64s, 128s, 256s, 512s, 1000s (max)
+	maxRetries = 15
+
+	// maxEndpointsPerSubset is the maximum number of endpoints that will be
+	// mirrored from an EndpointSubset.
+	maxEndpointsPerSubset = 1000
```

Comment:
This is actually distinct from the `maxEndpointsPerSlice` that can be configured with a corresponding `kube-controller-manager` flag and has the same default value of 100. This was an arbitrary number introduced to ensure we weren't mirroring huge numbers of endpoints. At a certain scale, it likely makes more sense for users of custom endpoints to be writing EndpointSlices instead of relying on this mirroring mechanism. This is a relatively arbitrary value that lines up with the max size for an EndpointSlice so we could theoretically achieve a 1:1 mapping between EndpointSubsets and EndpointSlices, but that's not actually what ended up happening here.

Reply:
I've cleaned this up so that `maxEndpointsPerSubset` is configurable as a controller flag and replaces the `maxEndpointsPerSlice` altogether.